internal/dcache/rpc/server/handler.go

```diff
@@ -47,43 +57,980 @@ var _ service.ChunkService = &ChunkServiceHandler{}
+		return rpc.NewResponseError(rpc.MVNotHostedByRV, fmt.Sprintf("MV %s is not hosted by RV %s", address.MvName, rvInfo.rvName))
+	}
+
+	return nil
+}
+
+// get the RVInfo from the RV name
+func (h *ChunkServiceHandler) getRVInfoFromRvName(rvName string) *rvInfo {
+	var rvInfo *rvInfo
+	for rvID, info := range h.rvIDMap {
+		common.Assert(info != nil, fmt.Sprintf("rvInfo nil for rvID %s", rvID))
+
+		if info.rvName == rvName {
+			rvInfo = info
+			break
+		}
+	}
+
+	return rvInfo
+}
+
+func (h *ChunkServiceHandler) createMVDirectory(path string) error {
+	log.Debug("ChunkServiceHandler::createMVDirectory: Creating MV directory %s", path)
+
+	if err := os.MkdirAll(path, 0700); err != nil && !os.IsExist(err) {
+		return fmt.Errorf("MkdirAll(%s) failed: %v", path, err)
+	}
+
+	return nil
 }
 
 func (h *ChunkServiceHandler) Hello(ctx context.Context, req *models.HelloRequest) (*models.HelloResponse, error) {
-	return nil, nil
+	if req == nil {
+		log.Err("ChunkServiceHandler::Hello: Received nil Hello request")
+		common.Assert(false, "received nil Hello request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil Hello request")
+	}
+
+	log.Debug("ChunkServiceHandler::Hello: Received Hello request: %+v", *req)
+
+	// // TODO:: integration: call cluster manager to get the node ID of this node
+	myNodeID := getNodeUUID()
+	common.Assert(req.ReceiverNodeID == myNodeID, "Received Hello RPC destined for another node", req.ReceiverNodeID, myNodeID)
+
+	// get all the RVs exported by this node
+	myRvList := make([]string, 0)
+	for _, info := range h.rvIDMap {
+		myRvList = append(myRvList, info.rvName)
+	}
+
+	return &models.HelloResponse{
+		ReceiverNodeID: myNodeID,
+		Time:           time.Now().UnixMicro(),
+		RV:             myRvList,
+		MV:             req.MV, // TODO:: discuss: how to fetch the MV list receiver shares with the sender
+	}, nil
 }
 
 func (h *ChunkServiceHandler) GetChunk(ctx context.Context, req *models.GetChunkRequest) (*models.GetChunkResponse, error) {
-	return nil, nil
+	if req == nil || req.Address == nil {
+		log.Err("ChunkServiceHandler::GetChunk: Received nil GetChunk request")
+		common.Assert(false, "received nil GetChunk request")
+		return nil, rpc.NewResponseError(rpc.InvalidRequest, "received nil GetChunk request")
+	}
+
+	log.Debug("ChunkServiceHandler::GetChunk: Received GetChunk request: %+v", *req)
+
+	// check if the chunk address is valid
+	err := h.checkValidChunkAddress(req.Address)
+	if err != nil {
+		log.Err("ChunkServiceHandler::GetChunk: Invalid chunk address [%s]", err.Error())
+		return nil, err
+	}
+
+	rvInfo := h.rvIDMap[req.Address.RvID]
+	mvInfo := rvInfo.getMVInfo(req.Address.MvName)
+
+	// Block the calling thread if this MV is currently quiesced
+	err = mvInfo.blockIOIfMVQuiesced()
+	common.Assert(err == nil, fmt.Sprintf("failed to block IO for MV %s", mvInfo.mvName))
```

Comment:
TODO: [Race] After blockIOIfMVQuiesced() decides to proceed and before we increment the ongoing IO count below, if quiesceIOsStart() is run it will succesfully mark the IOs as quiesced. This leads us to a state where IOs are running while they should be quiesced.

Reply:
Added TODO